sys/arm64/vmm/vmm_hyp.c

```diff
@@ -0,0 +1,802 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2021 Andrew Turner
+ *
+ * This work was supported by Innovate UK project 105694, "Digital Security
+ * by Design (DSbD) Technology Platform Prototype".
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/types.h>
+#include <sys/proc.h>
+
+#include <machine/armreg.h>
+
+#include "arm64.h"
+#include "hyp.h"
+
+struct hypctx;
+
+uint64_t vmm_hyp_enter(uint64_t, uintptr_t, uintptr_t, uintptr_t, uintptr_t,
+    uintptr_t, uintptr_t, uintptr_t);
+uint64_t vmm_enter_guest(struct hypctx *);
+
+static void
+vmm_hyp_reg_store(struct hypctx *hypctx, struct hyp *hyp, bool guest)
+{
+	uint64_t dfr0;
+
+	/* Store the guest VFP registers */
+	if (guest) {
+		/* Store the timer registers */
+		hypctx->vtimer_cpu.cntkctl_el1 = READ_SPECIALREG(cntkctl_el1);
+		hypctx->vtimer_cpu.virt_timer.cntx_cval_el0 =
+		    READ_SPECIALREG(cntv_cval_el0);
+		hypctx->vtimer_cpu.virt_timer.cntx_ctl_el0 =
+		    READ_SPECIALREG(cntv_ctl_el0);
+
+		/* Store the GICv3 registers */
+		hypctx->vgic_v3_regs.ich_eisr_el2 =
+		    READ_SPECIALREG(ich_eisr_el2);
+		hypctx->vgic_v3_regs.ich_elrsr_el2 =
+		    READ_SPECIALREG(ich_elrsr_el2);
+		hypctx->vgic_v3_regs.ich_hcr_el2 =
+		    READ_SPECIALREG(ich_hcr_el2);
+		hypctx->vgic_v3_regs.ich_misr_el2 =
+		    READ_SPECIALREG(ich_misr_el2);
+		hypctx->vgic_v3_regs.ich_vmcr_el2 =
+		    READ_SPECIALREG(ich_vmcr_el2);
+		switch (hypctx->vgic_v3_regs.ich_lr_num - 1) {
+#define	STORE_LR(x)					\
+	case x:						\
+		hypctx->vgic_v3_regs.ich_lr_el2[x] =	\
+		    READ_SPECIALREG(ich_lr ## x ##_el2)
```

Comment:
I find myself wanting these indented to align with the switch. I was confused when I hit the `default:` and the case wasn't aligned.